proof/refine/ARM/Schedule_R.thy

```diff
@@ -3368,5 +3317,840 @@ lemma possible_switch_to_corres:
+       apply (rule corres_add_noop_lhs)
+       apply (rule corres_split[OF _ threadSet_corres_noop])
+               apply (rule corres_split[OF _ setReprogramTimer_corres])
+                 apply simp
+                apply wpsimp
+               apply wpsimp
+              apply (clarsimp simp: tcb_relation_def)+
+        apply wpsimp+
+      apply (rule_tac P="valid_release_q and (\<lambda>s. release_queue s \<noteq> [] \<and> release_queue s = rq)"
+                  and P'="\<lambda>s'. ksReleaseQueue s' \<noteq> []"
+                   in corres_inst)
+      apply (rule corres_modify)
+      apply (fastforce intro: release_queue_modify_tl)
+     apply wpsimp+
+  apply (fastforce simp: st_tcb_at'_def)
+  done
+
+crunches tcb_release_remove
+  for valid_tcbs[wp]: valid_tcbs
+  (wp: crunch_wps update_valid_tcbs simp: crunch_simps)
+
+lemma tcbInReleaseQueue_update_valid_tcb'[wp]:
+  "threadSet (tcbInReleaseQueue_update (\<lambda>_. b)) tcbPtr \<lbrace>valid_tcb' tcb\<rbrace>"
+  apply (wpsimp wp: threadSet_valid_tcb')
+  apply (fastforce simp: valid_tcb'_def tcb_cte_cases_def valid_tcb_state'_def)
+  done
+
+lemma tcbInReleaseQueue_update_valid_tcbs'[wp]:
+  "threadSet (tcbInReleaseQueue_update (\<lambda>_. b)) tcbPtr \<lbrace>valid_tcbs'\<rbrace>"
+  apply (wpsimp wp: threadSet_valid_tcbs')
+  apply (fastforce simp: valid_tcb'_def tcb_cte_cases_def valid_tcb_state'_def)
+  done
+
+lemma tcbReleaseDequeue_valid_tcb'[wp]:
+  "tcbReleaseDequeue \<lbrace>valid_tcb' tcb\<rbrace>"
+  apply (clarsimp simp: tcbReleaseDequeue_def setReleaseQueue_def setReprogramTimer_def)
+  apply ((rule hoare_seq_ext_skip, wpsimp simp: update_valid_tcb') | wpsimp)+
+  done
+
+lemma tcbReleaseDequeue_valid_tcbs'[wp]:
+  "tcbReleaseDequeue \<lbrace>valid_tcbs'\<rbrace>"
+  apply (clarsimp simp: tcbReleaseDequeue_def setReleaseQueue_def setReprogramTimer_def)
+  apply ((rule hoare_seq_ext_skip, wpsimp simp: valid_tcbs'_def update_valid_tcb') | wpsimp)+
+  done
+
+lemma distinct_release_queue_cross_rel:
+  "cross_rel (pspace_aligned and pspace_distinct and valid_release_q)
+             (\<lambda>s'. distinct (ksReleaseQueue s'))"
+  apply (clarsimp simp: cross_rel_def)
+  apply (fastforce dest: state_relation_release_queue_relation
+                 intro!: tcb_at_cross
+                   simp: valid_release_q_def release_queue_relation_def vs_all_heap_simps
+                         obj_at_def is_tcb_def)
+  done
+
+lemma release_queue_nonempty_cross_rel:
+  "cross_rel (pspace_aligned and pspace_distinct and (\<lambda>s. release_queue s \<noteq> []))
+             (\<lambda>s'. ksReleaseQueue s' \<noteq> [])"
+  apply (clarsimp simp: cross_rel_def)
+   apply (fastforce dest: state_relation_release_queue_relation
+                  intro!: tcb_at_cross
+                    simp: valid_release_q_def release_queue_relation_def vs_all_heap_simps
+                          obj_at_def is_tcb_def)
+  done
+
+(* FIXME RT: move? *)
+lemma assert_drop:
```

Comment:
You'd usually want to either remove Q from the precondition or have it as implication. Putting it in as something to prove does not look useful (creates unnecessary work).

Reply:
Agreed! The default action is to use `assert_wp`, and if the implication is annoying to deal with then you can use `assert_inv` instead.